zarr/v3/codecs/transpose.py

```diff
@@ -32,82 +31,70 @@ class TransposeCodecMetadata:
             )
-            assert len(configuration.order) == len(set(configuration.order)), (
-                "There must not be duplicates in the `order` tuple. "
-                + f"Got: {configuration.order}"
+            assert len(self.order) == len(set(self.order)), (
+                "There must not be duplicates in the `order` tuple. " + f"Got: {self.order}"
             )
-            assert all(0 <= x < array_metadata.ndim for x in configuration.order), (
+            assert all(0 <= x < ndim for x in self.order), (
                 "All entries in the `order` tuple must be between 0 and "
-                + f"the number of dimensions in the array. Got: {configuration.order}"
+                + f"the number of dimensions in the array. Got: {self.order}"
             )
-            order = tuple(configuration.order)
+            order = tuple(self.order)
 
-        return cls(
-            array_metadata=array_metadata,
-            order=order,
-        )
+        if order != self.order:
+            return evolve(self, order=order)
+        return self
 
     @classmethod
     def get_metadata_class(cls) -> Type[TransposeCodecMetadata]:
         return TransposeCodecMetadata
 
-    def resolve_metadata(self) -> CoreArrayMetadata:
-        from zarr.v3.metadata import CoreArrayMetadata
+    def resolve_metadata(self, chunk_metadata: ChunkMetadata) -> ChunkMetadata:
+        from zarr.v3.metadata import ChunkMetadata
 
-        return CoreArrayMetadata(
-            shape=tuple(
-                self.array_metadata.shape[self.order[i]] for i in range(self.array_metadata.ndim)
-            ),
+        return ChunkMetadata(
             chunk_shape=tuple(
-                self.array_metadata.chunk_shape[self.order[i]]
-                for i in range(self.array_metadata.ndim)
+                chunk_metadata.chunk_shape[self.order[i]] for i in range(chunk_metadata.ndim)
             ),
-            data_type=self.array_metadata.data_type,
-            fill_value=self.array_metadata.fill_value,
-            runtime_configuration=self.array_metadata.runtime_configuration,
+            data_type=chunk_metadata.data_type,
+            fill_value=chunk_metadata.fill_value,
+            runtime_configuration=chunk_metadata.runtime_configuration,
         )
 
-    async def decode(
-        self,
-        chunk_array: np.ndarray,
-    ) -> np.ndarray:
-        inverse_order = [0 for _ in range(self.array_metadata.ndim)]
+    async def decode(self, chunk_array: np.ndarray, chunk_metadata: ChunkMetadata) -> np.ndarray:
+        inverse_order = [0 for _ in range(chunk_metadata.ndim)]
         for x, i in enumerate(self.order):
             inverse_order[x] = i
         chunk_array = chunk_array.transpose(inverse_order)
         return chunk_array
 
     async def encode(
-        self,
-        chunk_array: np.ndarray,
+        self, chunk_array: np.ndarray, chunk_metadata: ChunkMetadata
```

Comment:
is it important that this method operate on _chunks_, as opposed to _arrays_ (of which chunks are a subset)? If we want these functions to be agnostic to the semantics of chunking, and scoped more broadly to transforming arrays / bytes, then maybe renaming the `chunk_array` and `chunk_metadata` parameters to `array` and `array_metadata` would make sense?

Reply:
`transpose` is an `ArrayArrayCodec` that is why the arguments are called `chunk_array` in contrast to `chunk_bytes`.